src/coredump/coredump.c

```diff
@@ -1328,6 +1329,257 @@ static int gather_pid_metadata_from_procfs(struct iovec_wrapper *iovw, Context *
+        if (ERRNO_IS_NEG_XATTR_ABSENT(r))
+                return false;
+        if (r < 0)
+                return r;
+
+        return r > 0;
+}
+
+static int forward_coredump_to_container(Context *context) {
+        _cleanup_close_ int pidnsfd = -EBADF, mntnsfd = -EBADF, netnsfd = -EBADF, usernsfd = -EBADF, rootfd = -EBADF;
+        _cleanup_close_pair_ int pair[2] = PIPE_EBADF;
+        pid_t pid;
+        struct ucred ucred = {
+                .pid = context->pid,
+                .uid = context->uid,
+                .gid = context->gid,
+        };
+        int r;
+
+        r = namespace_get_leader(context->pid, NAMESPACE_PID, &pid);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get namespace leader: %m");
+
+        r = can_forward_coredump(pid);
+        if (r <= 0)
+                return log_debug_errno(r < 0 ? r : SYNTHETIC_ERRNO(ENOENT), "Cannot forward coredump: %m");
+
+        r = RET_NERRNO(socketpair(AF_UNIX, SOCK_DGRAM, 0, pair));
+        if (r < 0)
+                return log_debug_errno(r, "Failed to create socket pair: %m");
+
+        /* We need to translate the PID, UID, and GID of the crashing process
+         * to the container's namespaces. Do this by sending an SCM_CREDENTIALS
+         * message on a socket pair, and read the result when we join the
+         * container. The kernel will perform the translation for us. */
+        r = send_ucred_and_one_fd(pair[0], &ucred, STDIN_FILENO);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to send metadata to container: %m");
+
+        pair[0] = safe_close(pair[0]);
+
+        r = namespace_open(pid, &pidnsfd, &mntnsfd, &netnsfd, &usernsfd, &rootfd);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to join namespaces of PID " PID_FMT ": %m", pid);
+
+        r = namespace_fork("(sd-coredumpns)", "(sd-coredump)", NULL, 0,
+                           FORK_RESET_SIGNALS|FORK_DEATHSIG|FORK_WAIT,
+                           pidnsfd, mntnsfd, netnsfd, usernsfd, rootfd, NULL);
+        if (r == -EPROTO)
+                return log_debug_errno(r, "Failed to process coredump in container");
+        if (r < 0)
+                return log_debug_errno(r, "Failed to fork into namespaces of PID " PID_FMT ": %m", pid);
+        if (r == 0) {
+                _cleanup_(iovw_free_freep) struct iovec_wrapper *iovw = NULL;
+                _cleanup_close_ int input_fd = -EBADF;
+                char pidstr[DECIMAL_STR_MAX(pid_t)];
+                char uidstr[DECIMAL_STR_MAX(uid_t)];
+                char gidstr[DECIMAL_STR_MAX(gid_t)];
+                Context child_context = {};
+
+                if (laccess("/usr/lib/systemd/systemd-coredump", X_OK) < 0) {
+                        log_debug("Cannot find systemd-coredump, exiting.");
+                        _exit(EXIT_FAILURE);
+                }
+
+                r = receive_ucred_and_one_fd(pair[1], &ucred, &input_fd);
```

Comment:
there's no reason to send the input_fd via the socket btw, the child just gets the fd inherited anyway

Reply:
i think it would make sense to close `pair[1]` right after you read the datagram. not because it was necessary (it isn't), but simply to communicate to the reader that we now used the socket for what we wanted to use it for.